qemu/tests/hotplug_mem_share_discard_data.py

```diff
@@ -0,0 +1,87 @@
+from avocado.utils import process
+
+from virttest import error_context
+from virttest import env_process
+
+from virttest.utils_numeric import normalize_data_size
+from virttest.utils_test.qemu import MemoryHotplugTest
+from virttest.staging import utils_memory
+
+
+@error_context.context_aware
+def run(test, params, env):
+    """
+    Memory share and discard-data hotplug test
+    1) Setup hugepages
+    2) Mount /mnt/kvm_hugepage
+    3) Boot guest with one numa node
+    4) Hotplug 1G memory
+    5) Check memory
+    6) Unplug memory
+    7) Check memory
+    8) Check results for discard-data value
+
+    :param test: QEMU test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment
+    """
+
+    timeout = float(params.get("login_timeout", 240))
+
+    mem_dev = params.get('mem_devs')
+    size_mem = int(normalize_data_size(params['size_mem_%s' % mem_dev], 'K'))
+    total_hg_size = size_mem
+
+    target_mem = params.get("target_mems")
+    if params.get('backend_mem_%s' % target_mem) == 'memory-backend-file':
+        size_target_mem = int(normalize_data_size(params['size_mem_%s' % target_mem], 'K'))
+        total_hg_size += size_target_mem
+
+    hp_size = utils_memory.read_from_meminfo("Hugepagesize")
+    params['target_hugepages'] = int(total_hg_size // hp_size)
+    params['setup_hugepages'] = "yes"
+    params['not_preprocess'] = "no"
+
+    env_process.preprocess(test, params, env)
+
+    vm = env.get_vm(params["main_vm"])
+    vm.verify_alive()
+    session = vm.wait_for_login(timeout=timeout)
+
+    hotplug_test = MemoryHotplugTest(test, params, env)
+    hotplug_test.hotplug_memory(vm, target_mem)
+    hotplug_test.check_memory(vm)
+    hotplug_test.unplug_memory(vm, target_mem)
+    hotplug_test.check_memory(vm)
+
+    session.close()
+    error_context.context("Shutdown guest...", test.log.debug)
+    vm.destroy()
+
+    hp_total = int(utils_memory.read_from_meminfo("HugePages_Total"))
+    hp_free = int(utils_memory.read_from_meminfo('HugePages_Free'))
+
+    error_context.context("hp_total: %s" % str(hp_total), test.log.debug)
+    error_context.context("hp_free: %s" % str(hp_free), test.log.debug)
+
+    if not params.get_boolean("discard-data_plug1", True):
```

Comment:
Unfortunately, if you use params.get_boolean there must be a corresponding value in cfg.
like this:
`SKIP: Mandatory parameter 'mem-path_plug1' is missing. Check your cfg files for typos/mistakes`

Empty value if you use params.get cfg with no value

`if params.get("backend_mem_plug1") == "memory-backend-file":`

Reply:
Ok so you suggest just to only use: `if params.get("discard-data_plug1") == "no"` right?

Reply:
yes @yanan-fu do you have any suggestions?

Reply:
> yes @yanan-fu do you have any suggestions?

Then what about the final part (L83 - L87) ? This part needs to be done only `if params.get("backend_mem_plug1") == "memory-backend-file":`

Reply:
why?
(L83 - L87) is not in the judgment of if, so it will not affect

Reply:
Because at L83: `process.system("rm -rf %s" % params["mem-path_plug1"])`

**mem-path_plug1** parameter only exists at **backend-file** case so if only checking the discard-data this will be executed for backend-ram subcase too, leading to SKIP the case due to:
`SKIP: Mandatory parameter 'mem-path_plug1' is missing. Check your cfg files for typos/mistakes`

Reply:
I understand
I checked your previous code
https://github.com/autotest/tp-qemu/commit/f2fcb5059914b3bb18629f6ae5fdb4ecf0bf76ab

befor:
```
    if params.get("backend_mem_plug1") == "memory-backend-file":
        if not params.get_boolean("discard-data_plug1", True):

```
if you want to change:

`if params.get("backend_mem_plug1") == "memory-backend-file":`

It should be level with params.get("backend_mem_plug1")

Reply:
Ok so then it makes sense to keep the two if conditions right?

Reply:
```
    if params.get("backend_mem_plug1") == "memory-backend-file":
        if not params.get_boolean("discard-data_plug1", True):
        ......
        if hp_total != hp_free_after_delete:
            test.fail("Error, free hugepages doesn't match with total hugepages after deleting mem-path file")
```

change to:

```
    if params.get("backend_mem_plug1") == "no":
        ......
        if hp_total != hp_free_after_delete:
            test.fail("Error, free hugepages doesn't match with total hugepages after deleting mem-path file")
```

Reply:
> Unfortunately, if you use params.get_boolean there must be a corresponding value in cfg. like this: `SKIP: Mandatory parameter 'mem-path_plug1' is missing. Check your cfg files for typos/mistakes`
> 
> Empty value if you use params.get cfg with no value
> 
> `if params.get("backend_mem_plug1") == "memory-backend-file":`

Actually, `get_boolean` support no value in cfg,  and the error you hit is for `backend_mem_plug1` which use `get`,  not `discard_data-plug1`. 

The reason is in the pre command, check backend be removed, which cause the following step that remove backed_file be executed for all backend.
```
L67-L68
discard-data should only support backend_file mode, so we can reduce one judgment

if params.get("discard-data_plug1") == "no"
```

Reply:
> The reason is in the pre command, check backend be removed, which cause the following step that remove backed_file be executed for all backend.

But then what about L84 ? this deletion is mandatory and needs to be performed only for backend-file subcases

Reply:
@mcasquer  You are right, and we are in the same page. 
The comment above the is for the discussion with @zhenyzha before,  as remove the judgment of backend in the pre update, cause the Mandatory parameter missing error.
In short, the version f2fcb5059914b3bb18629f6ae5fdb4ecf0bf76ab is ok for this part :)